src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java

```diff
@@ -964,4 +895,83 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
+     * Average value is set to 0 if there are no assigned weights.
+     * e.g.
+     * percentageFrequencyAndAverageValue[subQuestionIndex][choiceIndex]
+     * -> is the percentage choiceIndex is chosen for subQuestionIndex
+     * percentageFrequencyAndAverageValue[subQuestionIndex][numOfRubricChoices]
+     * -> is the average weight of the responses for the given sub-question
+     */
+    private static class RubricStatistics {
+
+        int[][] responseFrequency;
+        float[][] percentageFrequencyAndAverage;
+
+        RubricStatistics(List<FeedbackResponseAttributes> responses, FeedbackRubricQuestionDetails questionDetails) {
+
+            int numOfRubricSubQuestions = questionDetails.numOfRubricSubQuestions;
+            int numOfRubricChoices = questionDetails.numOfRubricChoices;
+            int responseTotalIndex = numOfRubricChoices;
+
+            responseFrequency =
+                    new int[numOfRubricSubQuestions][numOfRubricChoices + 1];
+            percentageFrequencyAndAverage =
+                    new float[numOfRubricSubQuestions][numOfRubricChoices + 1];
+
+            // Count frequencies
+            for (FeedbackResponseAttributes response : responses) {
+                FeedbackRubricResponseDetails frd = (FeedbackRubricResponseDetails) response.getResponseDetails();
+                for (int i = 0; i < numOfRubricSubQuestions; i++) {
+                    int chosenChoice = frd.getAnswer(i);
+                    if (chosenChoice != -1) {
+                        responseFrequency[i][chosenChoice] += 1;
+                        responseFrequency[i][responseTotalIndex] += 1;
+                    }
+                }
+            }
+
+            // calculate percentage frequencies and average value
+            for (int i = 0; i < percentageFrequencyAndAverage.length; i++) {
+                int totalForSubQuestion = responseFrequency[i][responseTotalIndex];
+                //continue to next row if no response for this sub-question
+                if (totalForSubQuestion == 0) {
+                    continue;
+                }
+                // Divide responsesFrequency by totalForSubQuestion to get percentage
+                for (int j = 0; j < numOfRubricChoices; j++) {
+                    percentageFrequencyAndAverage[i][j] = (float) responseFrequency[i][j] / totalForSubQuestion;
+                }
+                // Calculate the average for each sub-question
+                if (questionDetails.hasAssignedWeights) {
+                    for (int j = 0; j < numOfRubricChoices; j++) {
+                        float choiceWeight =
+                                (float) (questionDetails.rubricWeights.get(j) * percentageFrequencyAndAverage[i][j]);
+                        percentageFrequencyAndAverage[i][numOfRubricChoices] += choiceWeight;
+                    }
+                }
+            }
```

Comment:
Chose not to further extract parts of the loop out as the code is relatively simply and the second inner loop depends on the first. There are not much benefits in further extracting (If second inner loop is extracted out as it is, readers have jump to another method to fully understand the code. If the second inner loop is extracted out with its own outer loop, the code would require the percentage frequencies to be computed first, introducing additional dependency.)